ffi/src/engine_funcs.rs

```diff
@@ -0,0 +1,131 @@
+use delta_kernel_ffi_macros::handle_descriptor;
+use tracing::debug;
+use url::Url;
+
+use crate::{
+    scan::SharedSchema, EngineData, ExternEngine, ExternResult, IntoExternResult,
+    KernelStringSlice, NullableCvoid, SharedExternEngine, TryFromStringSlice,
+};
+
+use super::handle::Handle;
+
+#[repr(C)]
+pub struct FileMeta {
+    pub path: KernelStringSlice,
+    pub last_modified: i64,
+    pub size: usize,
+}
+
+impl TryFrom<&FileMeta> for delta_kernel::FileMeta {
+    type Error = Error;
+
+    fn try_from(fm: &FileMeta) -> Result<Self, Error> {
+        let path = unsafe { String::try_from_slice(&fm.path) }?;
+        let location = Url::parse(&path)?;
+        Ok(delta_kernel::FileMeta {
+            location,
+            last_modified: fm.last_modified,
+            size: fm.size,
+        })
+    }
+}
+
+// Intentionally opaque to the engine.
+pub struct FileReadResultIterator {
+    // Box -> Wrap its unsized content this struct is fixed-size with thin pointers.
+    // Item = Box<dyn EngineData>, see above, Vec<bool> -> can become a KernelBoolSlice
+    data: FileDataReadResultIterator,
+
+    // Also keep a reference to the external engine for its error allocator.
+    // Parquet and Json handlers don't hold any reference to the tokio reactor, so the iterator
+    // terminates early if the last engine goes out of scope.
+    engine: Arc<dyn ExternEngine>,
+}
+
+#[handle_descriptor(target=FileReadResultIterator, mutable=true, sized=true)]
+pub struct MutFileReadResultIterator;
+
+impl Drop for FileReadResultIterator {
+    fn drop(&mut self) {
+        debug!("dropping FileReadResultIterator");
+    }
+}
+
+/// Call the engine back with the next `EngingeData` batch read by Parquet/Json handler
+/// # Safety
+///
+/// The iterator must be valid (returned by [`read_parquet_files`]) and not yet freed by
+/// [`free_read_result_iter`]. The visitor function pointer must be non-null.
+#[no_mangle]
+pub unsafe extern "C" fn read_result_next(
+    mut data: Handle<MutFileReadResultIterator>,
+    engine_context: NullableCvoid,
+    engine_visitor: extern "C" fn(engine_context: NullableCvoid, engine_data: Handle<EngineData>),
+) -> ExternResult<bool> {
+    let iter = unsafe { data.as_mut() };
+    read_result_next_impl(iter, engine_context, engine_visitor)
+        .into_extern_result(iter.engine.error_allocator())
+}
+
+fn read_result_next_impl(
+    iter: &mut FileReadResultIterator,
+    engine_context: NullableCvoid,
+    engine_visitor: extern "C" fn(engine_context: NullableCvoid, engine_data: Handle<EngineData>),
+) -> DeltaResult<bool> {
+    if let Some(data) = iter.data.next().transpose()? {
+        (engine_visitor)(engine_context, data.into());
+        // TODO: calling `into_raw` in the visitor causes this to segfault
+        //       perhaps the callback needs to indicate if it took ownership or not
+        // unsafe { BoxHandle::drop_handle(data_handle) };
```

Comment:
Yeah, so here's my thinking: This data is allocated by the engine (or at least by the `Engine::[parquet/json]_handler`). So it makes sense that it would be "owned" by the engine, and we wouldn't try and free it. Furthermore, exactly what the example program is doing is probably "normal". That is, you get called back with some data read from a file, your going to put it into the set of results and pass it along. You really _don't_ want this to be freed by the kernel (usually).

So what I've done is:
1. Make the type be `Handle<ExclusiveEngineData>`. This to me indicates that the engine owns this data, and that the kernel can't make any promises about it regarding thread safety.
2. Add a `free_engine_data` function for the case an engine actually does want the kernel to do the freeing
3. Add to the comment for this function that the engine owns this data and needs to call the above if it wants kernel to free things

This might get more complex if we need an `ExternEngineData`, but I think for that we'd have some kind of "into_raw" that engines could use to get back their data and kernel would free just the "rust bits" of it.

Reply:
Not sure I quite followed 2/? Doesn't engine _always_ need to call `free_engine_data` in order to clean up the rust bits? (but 100% agree that kernel should otherwise stay out of it)